domain/modelagent/service/service.go

```diff
@@ -499,6 +579,356 @@ func (s *Service) SetUnitReportedAgentVersion(
 	return nil
 }
 
+// UpgradeModelTargetAgentVersion is responsible for upgrading the target agent
+// version of the current model to latest version available. The version that is
+// upgraded to is returned.
+//
+// All agents that run on behalf of entities within the model will be entities
+// within the model will be eventually upgraded to the new version after this
+// call successfully returns.
+//
+// The following errors may be expected:
+// - [modelagenterrors.CannotUpgradeControllerModel] when the current model is
+// the model running the Juju controller.
+// - [modelagenterrors.ModelUpgradeBlocker] when their exists a blocker in the
+// model that prevents the model from being upgraded.
+func (s *Service) UpgradeModelTargetAgentVersion(
+	ctx context.Context,
+) (semversion.Number, error) {
+	ctx, span := trace.Start(ctx, trace.NameFromFunc())
+	defer span.End()
+
+	recommendedVersion := jujuversion.Current
```

Comment:
The 3.6 behaviour is to explicitly look up the target version of the controller model if the user didn't specify a version. If this controller agent has not yet upgraded then `jujuversion.Current` won't match the target version in the db which is what we want to use.

Reply:
Addressed in another comment on this PR.